### PR TITLE
fix: Added Apply/Get/List/Delete methods to project_metadata to BaseRegistry

### PR DESCRIPTION
**Pull request by EXPEbdodla**
# What this PR does / why we need it:
1. Adds Apply/Get/List/Delete methods to project_metadata to BaseRegistry
2. Made project as Optional in list_project_metadata method so we can use it retrieve all projects when project is None. Can be deleted in future if necessary. 
3. Renamed _maybe_init_project_metadata to apply_project_metadata. Registry Server will have use cases to create projects in future. 
4. I'll implement the methods to remote registry in future PRs

# Which issue(s) this PR fixes:

# Misc

### CODE REVIEW

File: sdk/python/feast/infra/registry/base_registry.py [hunk 1]
```diff
@@ -545,22 +545,72 @@ def list_validation_references(
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def apply_project_metadata(
```
Comment (tokoko) [inline sdk/python/feast/infra/registry/base_registry.py:549]: Wouldn't it make more sense for this to also accept `ProjectMetadata` object as an input similar to other registry resources? I know the object doesn't hold much information right now, but if we plan for it to be "updatable", an update should happen through this method, right?

**Comment by tokoko** [comment]
This probably means we will have to treat `ProjectMetadata` as another one of `FeastObject` types and should be added in [here](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feast_object.py#L25). Similarly for rbac to work, we need it [here](https://github.com/feast-dev/feast/blob/master/protos/feast/core/Permission.proto#L38) as well, but that can be deferred until handling in remote registry is implemented. @dmartinol 

Another point... if we treat `apply_project_metadata` as basically project creation and `delete_project_metadata` as deletion, along with all the contents (implementations seem to indicate this), wouldn't better name for this new feast object be `Project` rather than `ProjectMetadata`?

File: sdk/python/feast/infra/registry/base_registry.py (hunk 1 above)
Comment (EXPEbdodla) [reply sdk/python/feast/infra/registry/base_registry.py:549]: I was not envisioning this as a updatable. At this point its create only. Fields we have for this are UUID and LastUpdateTimestamp. UUID should be constant. LastUpdateTimestamp should be update when any changes happen to any objects associated to the project.

**Comment by EXPEbdodla** [comment]
> Another point... if we treat apply_project_metadata as basically project creation and delete_project_metadata as deletion, along with all the contents (implementations seem to indicate this), wouldn't better name for this new feast object be Project rather than ProjectMetadata?

@tokoko You are right. We can introduce the Project object with spec and metadata. In SQL Registry, currently it's stored in feast_metadata table, that will eventually of no use. It should be the case in other registries too. 

Do you recommend support for both Project and ProjectMetadata for a while? Or replace Project with ProjectMeatadata?

**Comment by EXPEbdodla** [comment]
> This probably means we will have to treat ProjectMetadata as another one of FeastObject types and should be added in [here](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feast_object.py#L25).

We won't treat this as an object that users define in Feature Definitions. We can populate project name from feature_store.yaml. At this point, I'm not aware of any use cases for this.

**Comment by tokoko** [comment]
> Defining Project object may be a case for RBAC. I have no idea about this at this point. Looking forward for the suggestions on this.

yup, that's what I think as well. A lot of rbac revolves around tagging, so it makes sense to make it possible to tag projects and consequently we will have to let users update those tags. Maybe we can make it so that users are able to define `Project` objects in the repo, but revert to auto-generating one for them if they don't do it.

@franciscojavierarceo @HaoXuAI Any thoughts/objections about introducing `Project` feast object and phasing out `ProjectMetadata` in favor of it? I think it makes a lot of sense especially considering rbac.

**Comment by dmartinol** [comment]
> This probably means we will have to treat ProjectMetadata as another one of FeastObject types and should be added in [here](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feast_object.py#L25). Similarly for rbac to work, we need it [here](https://github.com/feast-dev/feast/blob/master/protos/feast/core/Permission.proto#L38) as well, but that can be deferred until handling in remote registry is implemented. @dmartinol

Agreee, while developing the registry_server we'll also need to assert the required permissions.

**Comment by HaoXuAI** [comment]
No objection. Can we have multiple projects per repo where different project use different online/offline store setup?

**Comment by EXPEbdodla** [comment]
> No objection. Can we have multiple projects per repo where different project use different online/offline store setup?

@HaoXuAI We are discussing about the multiple Projects in the issue: https://github.com/feast-dev/feast/issues/4199. Project object can help us enable multiple projects per repo.

**Comment by tokoko** [comment]
@EXPEbdodla this is no longer relevant, right?

**Comment by franciscojavierarceo** [comment]
Is this PR still relevant? Thanks for doing this @EXPEbdodla and sorry for the complications here.

**Comment by EXPEbdodla** [comment]
Not relevant. I'll close it.

**Pull request closed without merging**